Exchange auth code for access token

robertbylight opened the pull request:
## Summary
This PR handles adding an `oauth/token` POST endpoint, and implements one to many relation between user and access token(one user to many access tokens).
- receives a post json body that includes; grant type, auth code, cliend id, and code verifier
- responds with an access token, expiration time and token type
- makes sure the authorization code can only be used once


## Related issue(s)

[Implement token exchange](https://trello.com/c/ZnMQY23F/5-implement-token-exchange)


> 
> Summary: Now that the authorize endpoint has been implemented, set up a User model so you can tie the user to the client. This is needed to store in the database what user gives permission to which app.
> 
> Definition of Done:
> 
> Develop a User model with  fields: email, first name, and last name.
> 
> Establish associations between User and oauth client
> 
> Ensure validations are in place for fields
> 
> Add unit tests.
> 

## Testing done

### Manual testing 
#### Set up:
- [ ] run `bundle install`
- [ ] if you've reviewed this project before, you'll need to run `rails db:drop`, and `rails db:create` before running `rails db:migrate` since this work changes the schema
- [ ] run migrate command `rails db:migrate`
- [ ] start up server `bin/rails server`
- [ ] open console `bin/rails console`
- [ ] create a client, copy and paste this code snippet into console. 
```
OauthClient.create(
          client_id: 'abc123',
          client_name: 'pr robert client',
          redirect_uri: 'https://www.prrobert.com/callback',
        )
```
you should get a response like this from the console
```
#<OauthClient:0x0000000119f59280
 id: 1,
 client_id: "abc123",
 client_name: "pr robert client",
 redirect_uri: "https://www.prrobert.com/callback",
 created_at: "2025-08-21 18:25:31.487443000 +0000",
 updated_at: "2025-08-21 18:25:31.487443000 +0000"> 
```
- [ ] create a user,  copy and paste this code snippet into console. 

```
    User.create!(
      email: 'robert@gmail.com',
      first_name: 'robert',
      last_name: 'rodriguez'
    )
```
you should get a response like this from the console
```
#<User:0x0000000119a59348
 id: 1,
 email: "[FILTERED]",
 first_name: "robert",
 last_name: "rodriguez",
 created_at: "2025-08-21 18:25:51.861504000 +0000",
 updated_at: "2025-08-21 18:25:51.861504000 +0000"> 
```

#### Successful request:
note: use https://tonyxu-io.github.io/pkce-generator/ to get a code challenge_value used in step 1 and code verifyer used in step 2
- [ ] Step 1: first get auth code from server, perform a GET request using endpoint `http://localhost:3000/oauth/authorize?client_id=abc123&response_type=code&redirect_uri=https://www.prrobert.com/callback&user_id=1&code_challenge=47DEQpj8HBSa-_TImW-5JCeuQeRkm5NMpJWZG3hSuFU&code_challenge_method=S256` and receive a response that should look like this but with a different code value 
```
{
    "redirect_url": "https://www.prrobert.com/callback?code=2ae47a4a7df52862a9233ec681f69a7a53fd7561cd79363386ce41ca7eab8597"
}
```
- [ ] Step 2: perform a POST request using endpoint `http://localhost:3000/oauth/token`, make sure you have `Content-Type` header as `application/json`, then copy this json and paste it into the request body. 
- 
copy auth code from response and replace the code shown in the request body snipped below, so from your response above you would only copy the value for code. ie from above response you would copy `2ae47a4a7df52862a9233ec681f69a7a53fd7561cd79363386ce41ca7eab8597`

request body:
```
{
  "grant_type": "authorization_code",
  "code": "47cc4418534f3e2b99afb086fa7fc36e36c1378addb81c25aff67107811d6047",
  "client_id": "abc123",
  "code_verifier": "KdiwRYqb5Ef0h-2BeUCqPJR-d91U2nqIMLEKE7Jebw4"
}
```
you should get a response similar to this( your expires in time may vary):
```
{
    "access_token": "7fbda7974b6b0fe0fff33bda1c469dad61344975aaf118b23f1d83af20d26aa0",
    "token_type": "Bearer",
    "expires_in": 3599
}
```

#### Bad requests
##### Note: because the auth code gets deleted even after failed requests, you'll have to repeat successful request step 1 to get a fresh code for each of these test cases 
- [ ] Perform the post request with a missing grant type, using endpoint `http://localhost:3000/oauth/token`
copy this snippet into your request body
```
{
  "code": "47cc4418534f3e2b99afb086fa7fc36e36c1378addb81c25aff67107811d6047",
  "client_id": "abc123",
  "code_verifier": "KdiwRYqb5Ef0h-2BeUCqPJR-d91U2nqIMLEKE7Jebw4"
}
```
 you should receive a 400 bad request response that looks like this
```
{
    "error": "grant_type must be authorization_code"
}
```
- [ ] Perform a POST request with the same auth code after getting a secessfull response from step 1 in the successful instructions above, use endpoint `http://localhost:3000/oauth/token`
copy this snippet into your request body(note: replace the code hash with auth code you got from the successful response received when hitting the `oauth/authorize` endpoint)
```
{
  "grant_type": "authorization_code",
  "code": "47cc4418534f3e2b99afb086fa7fc36e36c1378addb81c25aff67107811d6047",
  "client_id": "abc123",
  "code_verifier": "KdiwRYqb5Ef0h-2BeUCqPJR-d91U2nqIMLEKE7Jebw4"
}
```
you should receive a 400 response that looks like this

```
{
    "error": "Invalid or expired authorization code"
}
```
- [ ] Perform a POST request with a random auth code, using endpoint `http://localhost:3000/oauth/token`
copy this snippet into your request body(note: replace this code hash with auth code you got from the successful response received when hitting the `oauth/authorize` endpoint)
```
{
  "grant_type": "authorization_code",
  "code": "12312321ljlkjkljl",
  "client_id": "abc123",
  "code_verifier": "KdiwRYqb5Ef0h-2BeUCqPJR-d91U2nqIMLEKE7Jebw4"
}
```
you should receive a 400 response that looks like this

```
{
    "error": "Invalid or expired authorization code"
}
```

- [ ] Perform the post request with a missing client_id, using endpoint, `http://localhost:3000/oauth/token` 
copy this snippet into your request body
```
{
  "grant_type": "authorization_code",
  "code": "7b9e90c7858abb197cbbb61900483c7b17576269d692150a72789561457fd669",
  "code_verifier": "KdiwRYqb5Ef0h-2BeUCqPJR-d91U2nqIMLEKE7Jebw4"
}
```
 you should receive a 400 bad request response that looks like this
```
{
    "error": "Missing client_id"
}
```
- [ ] Perform the post request with an invalid client_id, using endpoint `http://localhost:3000/oauth/token`
copy this snippet into your request body
```
{
  "grant_type": "authorization_code",
  "code": "7b9e90c7858abb197cbbb61900483c7b17576269d692150a72789561457fd669",
  "client_id": "sdfdsf",
  "code_verifier": "KdiwRYqb5Ef0h-2BeUCqPJR-d91U2nqIMLEKE7Jebw4"
}
```
 you should receive a 400 bad request response that looks like this
```
{
    "error": "Invalid client_id"
}
```
- [ ] perform a POST request without the code_verifier using endpoint `http://localhost:3000/oauth/token`. 
request body:
```
{
  "grant_type": "authorization_code",
  "code": "47cc4418534f3e2b99afb086fa7fc36e36c1378addb81c25aff67107811d6047",
  "client_id": "abc123",
}
```
you should get a response similar to this:
```
{
    "error": "Missing code_verifier"
}
```
- [ ] perform a POST request with an invalid code_verifier using endpoint `http://localhost:3000/oauth/token`. 
request body:

```
{
  "grant_type": "authorization_code",
  "code": "02436faf0f9ea38d92990f6df23c4b4b386ac1e9710c6b5fc2d402faaa8a575c",
  "client_id": "abc123",
  "code_verifier": "abc"
}
```
you should get a response similar to this:
```
{
    "error": "Invalid code_verifier"
}
```


#### running tests using Rspec
open terminal and cd into cloned repo and run these commands
- [ ] test for models `rspec spec/models`
- [ ] test for controller `rspec spec/requests/oauth_controller_spec.rb`
- [ ] test for services `rspec spec/services`
- [ ] run all tests `rspec`

## Screenshots
n/a
## What areas of the site does it impact?
n/a
## Acceptance criteria

- [ ]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [ ]  No error nor warning in the console.
- [ ]  Events are being sent to the appropriate logging solution
- [ ]  Documentation has been updated (link to documentation)
- [ ]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature

## Requested Feedback
Im not too familiar with the conventions for Oauth, so please let me know if something in this PR deviates from convention even if its not necessary to request a change for it.

bramleyjl reviewed (changes requested):
Edit: I'm just now seeing your [PKCE PR](https://github.com/robertbylight/simple-oauth-server/pull/18) - rather than merge this code with the incorrect `client_secret` implementation & then followup with a change to implement PKCE you should just have a single (or sequential) PR implementing the PKCE changes to authorize & token. This can be done in smaller chunks like 1. update authorize & 2. update token but the PRs should be worked on one at a time without the intention of rewriting behavior later in subsequent PRs. Let me know if you'd like to pair on getting this rolling.

So right now what you have is a shared-secret authentication architecture as opposed to PKCE where the client passes a value for both authorization & token requests and the application evaluates them against each other. I added more details in my main comment on PKCE below and included a link to the auth0 docs on it as well as some spots in `vets-api` where SiS is using that architecture.

I had a few other cleanup comments as well but the PKCE implementation is the main part of it. As this implementation grows in complexity I'd start considering breaking off chunks of business logic into their own `service` files.

robertbylight commented:
> Edit: I'm just now seeing your [PKCE PR](https://github.com/robertbylight/simple-oauth-server/pull/18) - rather than merge this code with the incorrect `client_secret` implementation & then followup with a change to implement PKCE you should just have a single (or sequential) PR implementing the PKCE changes to authorize & token. This can be done in smaller chunks like 1. update authorize & 2. update token but the PRs should be worked on one at a time without the intention of rewriting behavior later in subsequent PRs. Let me know if you'd like to pair on getting this rolling.
> 
> So right now what you have is a shared-secret authentication architecture as opposed to PKCE where the client passes a value for both authorization & token requests and the application evaluates them against each other. I added more details in my main comment on PKCE below and included a link to the auth0 docs on it as well as some spots in `vets-api` where SiS is using that architecture.
> 
> I had a few other cleanup comments as well but the PKCE implementation is the main part of it. As this implementation grows in complexity I'd start considering breaking off chunks of business logic into their own `service` files.

@bramleyjl yeah that makes sense, the pkce and auth grant was not an original part of the work and added after this work was already done, so thats why the pkce pr came later than this one. But it makes sense to order it in the way you noted. I can do a pr to update the pkce and rebase this pr after that gets merged. So for now Ill hold off on this one then and touch base with it after the pkce work is in.

bramleyjl reviewed (changes requested):
Left a bunch of comments, some related to ruby stuff like accessing instance parameters through `attr_reader`, `attr_accessor`, etc.

~~For the access token you need sign and hand it off in `/token`, then validate your app's signature for authenticated requests - checkout the SiS `access_token_jwt_encoder/decoder` files for what that looks like.~~

You're making good progress! Ping me if you have any questions or want to go over some code snippets together.

bramleyjl reviewed (changes requested):
Almost there! I had two main comments that get repeated a couple of times across files.

1. You shouldn't need to pass or validate the `redirect_uri` as part of authentication or the token request - the application making those requests just needs to prove that they are the right client, and then we (and they) can trust the `redirect_uri` that's already saved in the configuration.
2. For RSpec tests it's usually best to define a `subject` at the top of the describe block for each method to replace calling the same `class.new` or `class.new.validate` or whatever. You can also define a name for it when you declare it so it looks better than "subject" in the tests, like `subject(:access_token) { AccessToken.new(...`.

robertbylight commented on spec/models/access_token_spec.rb:
```diff
@@ -1,26 +1,19 @@
+  let(:oauth_client) { valid_client }
+  let(:expires_at) { 1.hour.from_now }
+
+  subject(:access_token) { AccessToken.new(token:, oauth_client:, user:, expires_at:) }
```
gotcha, okay that makes sense, ill bag that into my next pr so i can just close this on, thanks!
